### DISCUSSION TITLE
Communication fails (irreversible) after restart

### DISCUSSION THREAD

**Discussion by CSH12** [WillCodeForCats/solaredge-modbus-multi#393] [General]
**Describe the bug**
After a restart of HA the communication to modbus fails. 

**Expected behavior**
Communication established also after restart

**Screenshots**
![image](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/97809992/7dce3541-3073-456a-a6fe-4d3bdcda9cd3)

**Debug Logs**
2023-08-08 10:02:05.166 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] solaredge_modbus_multi configuration: number_of_inverters=1, start_device_id=1, detect_meters=True, detect_batteries=True, keep_modbus_open=True, adv_storage_control=False, adv_site_limit_control=False, allow_battery_energy_reset=False, battery_rating_adjust=0, 
2023-08-08 10:02:05.166 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] coordinator timeout is 21.6
2023-08-08 10:02:05.174 WARNING (MainThread) [pymodbus.logging] Failed to connect [Errno 111] Connect call failed ('192.168.2.113', 1502)
2023-08-08 10:02:05.176 DEBUG (MainThread) [custom_components.solaredge_modbus_multi] Finished fetching SolarEdge Coordinator data in 0.010 seconds (success: False)
2023-08-08 10:02:05.419 WARNING (MainThread) [pymodbus.logging] Failed to connect [Errno 111] Connect call failed ('192.168.2.113', 1502)

**Home Assistant (please complete the following information):**
 - Home Assistant Core Version: Hassos 10.4, HA 2023.8.1 
 - solaredge-modbus-multi Version: v2.4.1
 - Installation Type: Supervised

**Additional context**
After deleting the inegration, restarting HA and reintegrating of solaredge modbus multi, the integration is working properly

My configuration is the following: 
![image](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/97809992/b64f4748-0030-46ee-92b5-ff78a91bccbc)
![image](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/97809992/515886ee-0b14-4c57-a641-c0f3e4f721f0)
![image](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/97809992/04102314-046a-4f47-baa3-52f7e33b0df0)
![image](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/97809992/3520c180-7ff4-449b-9e70-d3316826bf11)

**Comment by WillCodeForCats** [comment]
The "Connect call failed" error is from pymodbus, not this integration.

You can try rebooting your inverter or the server Home Assistant is running on.

**Comment by froggi1930** [comment]
same issue here. reboot of Inverter and HA did not change anything. No other instance is connecting to the Converter Modbus.

[pymodbus.logging] Failed to connect [Errno 111] Connect call failed.

If I read a register manually through the modbus integration I receive a value immediately.

Home Assistant 2023.8.1
Supervisor 2023.08.1
Operating System 10.4
Frontend 20230802.0 - latest

**Comment by WillCodeForCats** [comment]
I have not seen this problem in my HAOS test environments.

**Comment by WillCodeForCats** [comment]
Also everyone needs to upload their diagnostic files and debug logs.

**Comment by froggi1930** [comment]
> Also everyone needs to upload their diagnostic files and debug logs.

I am new to Home Assistant. What files do you need and where do I find them?

**Comment by WillCodeForCats** [comment]
Go to any inverter or meter or battery device info screen and there is a button that says "download diagnostics", it will download a file that can be uploaded here.

**Comment by froggi1930** [comment]
> Go to any inverter or meter or battery device info screen and there is a button that says "download diagnostics", it will download a file that can be uploaded here.

Thanks... I will set up a fresh installation and set up the SolardEdge again. In case I run into the same issues I will post the log files.

**Comment by froggi1930** [comment]
I have made a fresh installation and get the same error. If I read the modbus with the same settings in iobroker I do not fail.
I can not upload the debug file here... :-(

**Comment by WillCodeForCats** [comment]
Those logs are from a different integration, not this one.

**Comment by froggi1930** [comment]
Sorry... my mistake. Looks like I tried the wrong one. Here is the log for your solution.

[home-assistant_solaredge_modbus_multi_2023-08-10T15-46-08.541Z.log](https://github.com/WillCodeForCats/solaredge-modbus-multi/files/12314675/home-assistant_solaredge_modbus_multi_2023-08-10T15-46-08.541Z.log)

**Comment by WillCodeForCats** [comment]
I don't know what iobroker is, show debug logs from that and your inverter modbus config.

**Comment by WillCodeForCats** [comment]
Also you do know the inverter can only take one connection at a time, if whatever iobroker is connected nothing else can connect.

**Comment by WillCodeForCats** [comment]
Converting to discussion because no evidence this is an integration issue.

**Comment by froggi1930** [comment]
Iobroker is simular to Homeassistant. This instance is offline, so no other program tries to reach the inverter.
As I am new to Homeassistant I do not know what to do extra except installing your solution... Do I need to configure Modbus seperately, or only through your config?
I can reach the Inverter from my Laptop on port 1502 and also read the values.

**Reply by WillCodeForCats** [reply]
If you're able to read modbus through other devices there's no reason this shouldn't work if the IP and port are the same.

Otherwise I don't really have any ideas why it would say failed to connect. Since you've tried another integration where those logs showed also can't connect the only thing that tells me is that it's not specifically a problem with my integration.

**Comment by CSH12** [comment]
Hi all, 

Sorry for the delayed info, but had no time to respond this week earlier. 

I encountered the error again today as I restarted the Home assistant. 
Attached the requested diagnostic info and the log. If some more Infos are required, let me know

In my system there is no other device connecting to the modbus - just a single connection from home assistant.

To resolve the error I had to delete it again and reintegrate it.

[home-assistant_solaredge_modbus_multi_2023-08-08T08-06-37.746Z.log](https://github.com/WillCodeForCats/solaredge-modbus-multi/files/12325048/home-assistant_solaredge_modbus_multi_2023-08-08T08-06-37.746Z.log)
[config_entry-solaredge_modbus_multi-e3d0a9ca86e2ff5a85c0a45012c00feb.json.txt](https://github.com/WillCodeForCats/solaredge-modbus-multi/files/12325051/config_entry-solaredge_modbus_multi-e3d0a9ca86e2ff5a85c0a45012c00feb.json.txt)


There is one more issue that is happening sometime. Some individual information (B1 Energy Export and Import, or Solar Production) are unavailable. But just some of them, all other ones are stable. So I don't know why. If I reload the integration it is working properly - strange.

**Reply by WillCodeForCats** [reply]
I don't really have any ideas at the moment why it would say failed to connect.

**Reply by CSH12** [reply]
would it make sense to configure the native modbus integration in HA, to log the raw information? So that you/someone is able to compare data?

**Reply by WillCodeForCats** [reply]
The native modbus integration is also based on pymodbus, the same library I use here. There shouldn't be a difference.

**Reply by CSH12** [reply]
any other way to generate a raw log without special equipment?

**Reply by CSH12** [reply]
and actually just one entity is missing again 
![image](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/97809992/66c1d5d4-f750-4b7e-86f4-fd158a8e1294)

is this maybe some timeout detection related to the information is not getting update within a tiemespan or sth.? 
seems that the information that are not updated frequently (night in germany, no solar production change) are detected as unavailable

**Reply by WillCodeForCats** [reply]
Night shouldn't change anything, mine don't change at night and just stay at the same value.

**Reply by WillCodeForCats** [reply]
What entity is that going unavailable? These are `sensor.solaredge_i1_ac_energy_kwh` and `sensor.solaredge_i1_ac_energy_kwh` on my system:

<img width="590" alt="Screenshot 2023-08-11 at 3 34 50 PM" src="https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/48533968/113e544e-a00e-414b-92b3-4fd1e166f5b5">
<img width="593" alt="Screenshot 2023-08-11 at 3 35 11 PM" src="https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/48533968/29c91eba-3bd2-40bd-9072-f1f8ccc01a14">

**Reply by CSH12** [reply]
just as an idea...
this is the only hint that I recognized that the not changing data are set to unavailable. 
It seems to be all time the same ones: solar production or B1 Energy Import, Export

**Reply by WillCodeForCats** [reply]
B1 import and export energy are a known issue that they "go backwards" when they should be increasing only: [Known-Issues#battery-energy](https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki/Known-Issues#battery-energy)

**Comment by john8329** [comment]
I'm having the same problem. The inverter becomes unreachable after restarting HASS. The only way to fix it is restarting the inverter. I presume it thinks there's an open connection and refuses any new one, remaining stuck in this state. Perhaps the integration is not closing the modbus connection properly on shutdown?

Here's the diags. Thanks for the component btw.
[home-assistant_solaredge_modbus_multi_2023-09-11T19-12-08.957Z.log](https://github.com/WillCodeForCats/solaredge-modbus-multi/files/12579582/home-assistant_solaredge_modbus_multi_2023-09-11T19-12-08.957Z.log)

**Reply by WillCodeForCats** [reply]
SolarEdge inverters are supposed to time out a connection after 2 minutes of inactivity.

The integration has a shutdown function that gets called when it's unloaded.

**Reply by john8329** [reply]
...and yet.
This problem seems very reproducible, the inverter is recent and updated, and it's always been like this. If I can, I'm willing to help to troubleshoot it.

**Reply by WillCodeForCats** [reply]
I don't see this issue on my system, so I can't reproduce it.

**Reply by john8329** [reply]
It is here unfortunately. I'm posting this because it looks like it's not an isolated issue, and since I have the time and reproducibility, we might be able to solve it.

No particular hurry of course, if you think it's not important, I'll live with having to unplug the inverter every time I restart HA.

**Reply by WillCodeForCats** [reply]
Practically speaking if I can't reproduce it or observe it, and there's no evidence in logs what or where the problem could be, there's no starting point to actually do anything about it.

Whether or not I think it's important is irrelevant - there is nothing for me to work with.

**Reply by WillCodeForCats** [reply]
Specifically in your logs you have:

`Modbus Error: [Connection] Connection lost during request`

That means it connected to the inverter and sent it a request, but the inverter dropped the connection somewhere in the process of receiving or sending data. So what the integration will do is issue a disconnect and setup will try again (wait time between setup retries is determined by Home Assistant). I don't know what other alternate sequence there would be to test or try with the information given.

At this point the only thing I recommend would be to disable the "Keep Modbus Open" setting and keep the polling time greater than 10 seconds.

**Reply by john8329** [reply]
Worth a shot, I'll try. My polling time is 5 seconds. Also I'll try keeping HA off for 5min, then turning it on again, if it fails to reconnect after that time it means the inverter is likely stuck in an invalid state, which could be caused by any reason, but I suspect the disconnect sequence.

I'll post updates here if I find anything.
Thanks

**Reply by WillCodeForCats** [reply]
Propose what should be changed if you think the sequence is incorrect.

**Reply by WillCodeForCats** [reply]
This is the shutdown function if you suspect it is out of sequence:

```
    async def shutdown(self) -> None:
        """Shut down the hub and disconnect."""
        async with self._lock:
            self.online = False
            self.disconnect()
            self._client = None

    def disconnect(self) -> None:
        """Disconnect from inverter."""
        if self._client is not None:
            self._client.close()
```

The only restriction is that `self.disconnect()` has to come before `self._client = None` otherwise `self._client.close()` does not exist.

**Reply by john8329** [reply]
I confirm it's the inverter becoming stuck. Even if I keep HA off for an hour, it fails reconnecting, and I have to reboot it. Hope this info can help somebody out there.

**Reply by john8329** [reply]
Also wait. Isn't this an async client? I'm not extremely practical with async python, but shouldn't you `await` the `close()` call when disconnecting? Got it from [here](https://pymodbus.readthedocs.io/en/v3.1.1/source/library/pymodbus.client.html)

**Reply by WillCodeForCats** [reply]
Close is not an async call since PR https://github.com/pymodbus-dev/pymodbus/pull/1492 starting with pymodbus v3.3.0
